changefeedccl: database-level changefeeds: tableset watcher

asg0451 opened the pull request:
This is a new package that implements a watcher
for database-level changefeeds. It is used to
watch for changes to the set of watched tables, so
that the feed coordinator may react to them.

Fixes: #148836
Fixes: #148829

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/149710)
<!-- Reviewable:end -->

asg0451 commented:
i will squash these into one commit when it's no longer a draft, but i'd appreciate a review in advance of that. thanks!

aerfrei commented on pkg/ccl/changefeedccl/tableset/watcher.go:
```diff
@@ -0,0 +1,567 @@
+
+	id      int64
+	execCfg *sql.ExecutorConfig
+	mon     *mon.BytesMonitor
```
My sense is that we don't think that this component should be a big consumer of memory and that this shouldn't be a bottleneck, is that right? What would happen if the watcher did start to use a lot memory?

aerfrei commented on pkg/ccl/changefeedccl/tableset/watcher.go:
```diff
@@ -0,0 +1,567 @@
+	upToIdx, found := slices.BinarySearchFunc(w.state.tableDiffs, upTo, func(diff TableDiff, ts hlc.Timestamp) int {
+		return diff.AsOf.Compare(ts)
+	})
+	if found && upToIdx > 0 {
```
I'm trying to understand why we decrement the index in this case. If we ask for the tableDiffs that have happened up to time upTo, we add a waiter and wait for it to finish. Then once it does, we find its id in the tableDiffs. It's possibly not there anymore (if a pop call with a later upTo finished first?). In either case, we want to remove all the tableDiffs that have happened at or before that time.

If we find a tableDiff at exactly that timestamp (at index upToIdx)
	diffs := w.state.tableDiffs[:upToIdx]
	w.state.tableDiffs = w.state.tableDiffs[upToIdx:]

diffs would not include that tableDiff even without the decrement (if I'm not having an off by one moment). That seems right because of that thing we were discussing a while back where a resolved ts of x does not mean that events at x were emitted. It's a guarantee about events strictly before x. But then why do we decrement on top of that? And why does it depend on if upToIdx is greater than 0?

asg0451 replied:
thats a good question and its entirely possible there's an off by 1 error here. i'll think about this some more

KeithCh commented on pkg/ccl/changefeedccl/tableset/watcher.go:
```diff
@@ -0,0 +1,567 @@
+
+// TableSetAt returns the tableset at the given timestamp. This is just a
+// convenience method and does not mutate the watcher.
+func (w *Watcher) TableSetAt(ctx context.Context, at hlc.Timestamp) (TableSet, error) {
```
Ditto on this function, its not used at all in this PR.

andyyang890 commented on pkg/ccl/changefeedccl/tableset/watcher.go:
```diff
@@ -0,0 +1,567 @@
+			select {
+			// Buffer incoming tables between resolved messages.
+			case diff := <-incomingTableDiffs:
+				if diff.AsOf.Less(curResolved) {
```
I'm not super confident because I don't know the rangefeed API that well, but I thought a checkpoint was supposed to be a guarantee that it won't emit any events with lower timestamps https://github.com/cockroachdb/cockroach/blob/2b4bb44a1ca27caff19e694b67f55ad3e5137769/pkg/kv/kvpb/api.proto#L3299-L3302

asg0451 commented on pkg/ccl/changefeedccl/tableset/watcher_test.go:
```diff
@@ -0,0 +1,295 @@
+	dbID := getDatabaseID(t, ctx, &execCfg, "defaultdb")
+	filter := Filter{
+		DatabaseID:    dbID,
+		ExcludeTables: []string{"exclude_me", "exclude_me_also"},
```
the options are exclusive, but i'll add some test coverage for IncludeTables too

asg0451 commented on pkg/ccl/changefeedccl/tableset/watcher.go:
```diff
@@ -0,0 +1,567 @@
+	// TODO: we technically dont need to wait for resolved here if there already
+	// are diffs buffered > upTo. But this seems tricky to implement rn.
+
+	diffs, err := w.Pop(ctx, upTo)
```
i don't think that's necessary... why?

andyyang890 replied:
Hmm I guess it's not clear to me how this function will be used, but I'm worried about possibly losing events. I suppose the actual events aren't useful sometimes? I think also the function name doesn't directly imply that events will be popped.

Maybe instead `Pop` could record the latest pop time and make sure that never regresses? If we want to be extra paranoid, it could accept both a start and end time to make sure there are no gaps.

Also, if the consumer is expected to always call this function instead of `Pop`, perhaps `Pop` should be unexported?

asg0451 replied:
the usage will be like 
```
for tick {
  if !watcher.UnchangedUpTo(time.Now()) {
     return replanErr
   }
}
```

how do you see events being lost here?

andyyang890 replied:
Yeah I guess that's fine, I was worried it might be used like
```
now := time.Now()
if !watcher.UnchangedUpTo(now) {
    events, err := watcher.Pop(now)
    // do something with events
}
```

aerfrei replied:
If we don't want people to use it by calling .Pop directly, does that mean Pop shouldn't be exported as Andy was suggesting?

asg0451 committed:
changefeedccl: database-level changefeeds: tableset watcher

This is a new package that implements a watcher
for database-level changefeeds. It is used to
watch for changes to the set of watched tables, so
that the feed coordinator may react to them.

Fixes: #148836
Fixes: #148829

Release note: None

aerfrei commented on pkg/ccl/changefeedccl/tableset/watcher.go:
```diff
@@ -0,0 +1,534 @@
+
+// Pop returns the table diffs between the last call to Pop() (or the initial
+// timestamp) and the given timestamp [inclusive, exclusive).
+// Users of this package will usually call `Pop` instead, but this is exposed for testing.
```
This is a typo I assume, ~Pop~ -> UnchangedUpTo

asg0451 commented on pkg/ccl/changefeedccl/tableset/watcher.go:
```diff
@@ -0,0 +1,534 @@
+				log.Infof(ctx, "onValue: %s; prev: %s", table, prevTable)
+			}
+
+			if !(w.filter.includes(table) || (kv.PrevValue.IsPresent() && w.filter.includes(prevTable))) {
```
i know this is common advice but i really dont think the demorganification of this is easier to understand; in fact i think it's harder. do you have a proposal?

andyyang890 replied:
ig when I read if statements, I like being able to short-circuit once I know the condition is definitely going to be true or false, but it's harder to do that when it's not DeMorgan-fied because then you basically have to do the DeMorgan in your head

if we apply DeMorgan's the condition becomes this:
```
!w.filter.includes(table) && (!kv.PrevValue.IsPresent() || !w.filter.includes(prevTable))
```

asg0451 commented on pkg/ccl/changefeedccl/tableset/watcher.go:
```diff
@@ -0,0 +1,534 @@
+type Filter struct {
+	// DatabaseID is the ID of the database to watch. It is mandatory.
+	DatabaseID    descpb.ID
+	IncludeTables map[string]struct{}
+	ExcludeTables map[string]struct{}
```
I think these will be replaced by whatever @KeithCh implements for this work, so I don't think it's important to do this here and now.

asg0451 replied:
will add a note

andyyang890 commented on pkg/ccl/changefeedccl/tableset/watcher.go:
```diff
@@ -0,0 +1,536 @@
+	execCfg *sql.ExecutorConfig
+	mon     *mon.BytesMonitor
+	acc     mon.BoundAccount // set on Start
+	started bool
```
maybe `started` should go in `mu` to protect against concurrent starts?

asg0451 replied:
ehhhhhhhh

andyyang890 commented on pkg/ccl/changefeedccl/tableset/watcher.go:
```diff
@@ -0,0 +1,536 @@
+}
+
+// PopUnchangedUpTo returns true if the tableset is unchanged between when it
+// (or Pop()) was last called (or the initialTS) and the given timestamp
```
nit: remove comments that reference no longer existent `Pop` function
